src/box/box.cc

```diff
@@ -1835,6 +1841,8 @@ box_promote_qsync(void)
 		diag_set(ClientError, ER_NOT_LEADER, raft->leader);
 		return -1;
 	}
+	if (box_raft_wait_term_broadcasted() != 0)
+		return -1;
 	box_issue_promote(txn_limbo.owner_id, wait_lsn);
```

Comment:
If the problem is only with promote/demote, can this waiting be done right inside of `box_issue_promote/demote()`?

Reply:
It can, the only problem is there is no return value from box_issue_promote/demote. Should I move it there anyway?

Reply:
Your function can only fail if the fiber is cancelled. But `box_issue_promote/demote()` are writing to disk. They are non-cancellable during that procedure. It means that you could make `box_raft_wait_term_broadcasted()` never failing and wait regardless of whether the fiber is cancelled. And it won't matter in case you will do it right before the WAL write. I don't think it would cause any problems. Lets try?